gap/oper.gi

```diff
@@ -2005,42 +2005,82 @@ end);
 InstallMethod(VerticesReachableFrom, "for a digraph and a vertex",
 [IsDigraph, IsPosInt],
 function(D, root)
-  local N, index, current, succ, visited, prev, n, i, parent,
-  have_visited_root;
+  local N;
   N := DigraphNrVertices(D);
+
   if 0 = root or root > N then
     ErrorNoReturn("the 2nd argument (root) is not a vertex of the 1st ",
                   "argument (a digraph)");
   fi;
-  index := ListWithIdenticalEntries(N, 0);
-  have_visited_root := false;
-  index[root] := 1;
-  current := root;
-  succ := OutNeighbours(D);
-  visited := [];
-  parent := [];
-  parent[root] := fail;
-  repeat
-    prev := current;
-    for i in [index[current] .. Length(succ[current])] do
-      n := succ[current][i];
-      if n = root and not have_visited_root then
-         Add(visited, root);
-         have_visited_root := true;
-      elif index[n] = 0 then
-        Add(visited, n);
-          parent[n] := current;
-          index[current] := i + 1;
-          current := n;
-          index[current] := 1;
-          break;
+
+  return VerticesReachableFrom(D, [root]);
+end);
+
+InstallMethod(VerticesReachableFrom, "for a digraph and a list of vertices",
+[IsDigraph, IsList],
+function(D, roots)
+  local N, index, visited, queue_tail, queue,
+  root, element, neighbour, graph_out_neighbors, node_neighbours;
+
+  N := DigraphNrVertices(D);
+  visited := BlistList([1 .. N], []);
+
+  for root in roots do
+    if 0 = root or root > N then
+      ErrorNoReturn("an element of the 2nd argument ",
+                    "(roots) is not a vertex of the 1st ",
+                    "argument (a digraph)");
+    fi;
+  od;
+
+  graph_out_neighbors := OutNeighbors(D);
+  queue := EmptyPlist(N);
+  queue_tail := 0;
+  for root in roots do
+    queue_tail := queue_tail + 1;
+    queue[queue_tail] := root;
```

Comment:
I'm not sure I see why you are doing this here, why not just do `queue := roots`? Or are you trying to accommodate the situation when `roots` has holes (i.e. `roots = [1,, 2]`)? If the latter, then I think the better approach would be to just require that `roots` satisfies `IsDenseList` and to error out if it does not.

Reply:
The idea here was for efficiency, queue := roots would create a list that would need resizing, which would (I think) have some time-cost associated with resizing the list, creating the list like this means we don't have to resize the array every time which I thought would save time, correct me if I'm wrong :)